gossiper: failure_detector_loop_for_node:  abort send_gossip_echo using abort_source

bhalevy opened the pull request:
Currently send_gossip_echo has a 22 seconds timeout
during which _abort_source is ignored.

Use a function-local abort_source to abort
send_gossip_echo either on timeout or if
_abort_source requested abort, and co_return in
the latter case.

avikivity commented on message/rpc_protocol_impl.hh:
```diff
@@ -180,24 +180,30 @@ auto send_message_cancellable(messaging_service* ms, messaging_verb verb, msg_ad
+    auto& c_ref = st->c;
+    st->sub = as.subscribe([st] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+        st->c.cancel();
+        st->ex = opt_ex.value_or(std::make_exception_ptr(abort_requested_exception{}));
```
A constructor in state would allow you to avoid optimized_optional. It's cosmetic, but one less thing to think about.

avikivity commented on message/rpc_protocol_impl.hh:
```diff
@@ -180,24 +180,30 @@ auto send_message_cancellable(messaging_service* ms, messaging_verb verb, msg_ad
+        st->ex = opt_ex.value_or(std::make_exception_ptr(abort_requested_exception{}));
     });
-    if (!sub) {
+    if (!st->sub) {
```
Doesn't the meaning change? Now it's checking the optimized_optional rather than sub itself.

bhalevy replied:
No, it was optimized_optional before too.
That's what subscribe returns.
See https://github.com/scylladb/seastar/blob/502f1422d6e3958237eadcada534fe46a734c8b3/include/seastar/core/abort_source.hh#L134

avikivity commented on message/rpc_protocol_impl.hh:
```diff
@@ -180,24 +180,30 @@ auto send_message_cancellable(messaging_service* ms, messaging_verb verb, msg_ad
-            // Translate low-level canceled_error into high-level abort_requested_exception.
-            return futurator::make_exception_future(abort_requested_exception{});
+            // Translate low-level canceled_error into high-level (typically abort_requested_exception).
+            return futurator::make_exception_future(std::move(st->ex));
```
Isn't this dangerous? If multiple rpcs are cancelled, only one will see st->ex.

bhalevy replied:
Oops, you're right.  We're not returning after the first faiure.

bhalevy replied:
Taking my words back.  `st` is a unique_ptr that's destroyed along with this continuation, and with `st->ex`.
Each rpc message allocates its own state and its subscription on the shared abort_source propagates this exception via the state.

avikivity replied:
Okay. But let's make sure we're not adding an allocation to every read or write.

bhalevy replied:
we're not

avikivity commented on gms/gossiper.cc:
```diff
@@ -762,16 +762,28 @@ future<> gossiper::failure_detector_loop_for_node(gms::inet_address node, int64_
+        tmr.cancel();
+        as.request_abort_ex(abort_requested_exception());
+    });
     while (is_enabled()) {
```
Can't we use _abort_source directly?

avikivity replied:
I think I see a reason why not, but the code doesn't consider it. If the timer ever fires, `as` becomes poisoned. We need a new one every iteration.

bhalevy replied:
Right again.

bhalevy replied:
Hmm, theoretically gossip can be stopped gracefully without requesting abort.
But I'll add a check like in https://github.com/scylladb/scylladb/blob/563998b69a1cb2db4e4dba7eb9f9703204588ffd/gms/gossiper.cc#L809

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3476/

bhalevy commented:
In v2 ([17f6005](https://github.com/scylladb/scylladb/commit/17f6005e93cf4ccd978d37edd7f8f962d1f2db50)):
- messaging: send_message_cancellable: allow passing exception via abort_source
  - subscribe in state ctor
  - try/catch exceptions around `as.check()` and possible allocation failures
- gossiper: failure_detector_loop_for_node: abort send_gossip_echo using abort_source
  - create abort_source in iteration body
  - timer and subscription are still created once and use a pointer to the loop abort_source
  - the external pointer to abort_source is engaged and nullified in the loop around the rpc call
  - protect against double abort

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3492/

bhalevy commented:
@avikivity ping, please re-review

bhalevy commented:
In v3 ([8b927e7e](https://github.com/scylladb/scylladb/commit/8b927e7efb0b8c10c8bed3b423efcef8dd1cae21))
- rebased + retested
- simplified PATCH 1/2
  - c_ref was unused
- added: "gossiper: failure_detector_loop_for_node: check if abort_requested in loop condition"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3897/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3899/

bhalevy commented:
@avikivity ping review

avikivity commented on message/rpc_protocol_impl.hh:
```diff
@@ -177,32 +177,44 @@ auto send_message_cancellable(messaging_service* ms, messaging_verb verb, msg_ad
+        explicit state(abort_source& as) {
+            as.check();
+            sub = as.subscribe([this] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+                c.cancel();
```
This adds an allocation. Do we have data-plane callers using this?

The data plane should likely manage its about sources on its own. This is convenient for the control plane.

bhalevy replied:
No, it's called only from `send_gossip_echo` and `send_schema_check`.

bhalevy replied:
Where is the extra allocation by the way?
We currently have `std::make_unique<seastar::rpc::cancellable>()`
and this changes it `std::make_unique<state>(as)`

avikivity commented on gms/gossiper.cc:
```diff
@@ -762,16 +762,38 @@ future<> gossiper::failure_detector_loop_for_node(gms::inet_address node, int64_
+        if (asp && !asp->abort_requested()) {
+            asp->request_abort_ex(abort_requested_exception());
+        }
+    });
```
This isn't the intent of abort_source. It's meant to be a long-term object that manages shutdowns, not something you create for every request.

bhalevy replied:
does it really matter?
it's so light weight that it can be used efficiently per request, certainly on the control plain.

avikivity commented:
I don't understand the goal of the series.

bhalevy commented:
To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.

avikivity commented:
> To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.

Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?

bhalevy commented:
> > To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.
> 
> Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?

Since we want the message to be abortable either by the timeout or the abort_source, whichever comes first.
We have `send_message_timeout` that is not abortable by an abort_source, and
we have send_message_cancellable that is abortable by an abort_source but supports no timeout.
Using an ad-hoc abort_source like in this case allows as to chain the abort request from the upper-level abort_source, or to yank it on timeout.

gleb-cloudius commented:
> > > To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.
> > 
> > 
> > Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?
> 
> Since we want the message to be abortable either by the timeout or the abort_source, whichever comes first. We have `send_message_timeout` that is not abortable by an abort_source, and we have send_message_cancellable that is abortable by an abort_source but supports no timeout. Using an ad-hoc abort_source like in this case allows as to chain the abort request from the upper-level abort_source, or to yank it on timeout.

We have `abort_on_expiry` for that. Can you use it here?

bhalevy commented:
> > > > To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.
> > > 
> > > 
> > > Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?
> > 
> > 
> > Since we want the message to be abortable either by the timeout or the abort_source, whichever comes first. We have `send_message_timeout` that is not abortable by an abort_source, and we have send_message_cancellable that is abortable by an abort_source but supports no timeout. Using an ad-hoc abort_source like in this case allows as to chain the abort request from the upper-level abort_source, or to yank it on timeout.
> 
> We have `abort_on_expiry` for that. Can you use it here?

Ah, I wasn't aware of it.
It could do the job if extended to subscribe on an external abort source.

gleb-cloudius commented:
> > > > > To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.
> > > > 
> > > > 
> > > > Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?
> > > 
> > > 
> > > Since we want the message to be abortable either by the timeout or the abort_source, whichever comes first. We have `send_message_timeout` that is not abortable by an abort_source, and we have send_message_cancellable that is abortable by an abort_source but supports no timeout. Using an ad-hoc abort_source like in this case allows as to chain the abort request from the upper-level abort_source, or to yank it on timeout.
> > 
> > 
> > We have `abort_on_expiry` for that. Can you use it here?
> 
> Ah, I wasn't aware of it. It could do the job if extended to subscribe on an external abort source.

May be we should have a wait to subscribe two two different abort sources instead of extending the primitive. something like `subscribe(or(as1, as2))` and `subscribe(and(as1, as2))` depending on if we want to abort if one of them is aborted or all of them.

bhalevy commented:
> > > > > > To save up to 22 seconds delay when shutting down gossip if it happens to wait in send_gossip_echo for timeout.
> > > > > 
> > > > > 
> > > > > Then why use a new abort_source? Can't we use the abort_source that shuts down gossip in the first place?
> > > > 
> > > > 
> > > > Since we want the message to be abortable either by the timeout or the abort_source, whichever comes first. We have `send_message_timeout` that is not abortable by an abort_source, and we have send_message_cancellable that is abortable by an abort_source but supports no timeout. Using an ad-hoc abort_source like in this case allows as to chain the abort request from the upper-level abort_source, or to yank it on timeout.
> > > 
> > > 
> > > We have `abort_on_expiry` for that. Can you use it here?
> > 
> > 
> > Ah, I wasn't aware of it. It could do the job if extended to subscribe on an external abort source.
> 
> May be we should have a wait to subscribe two two different abort sources instead of extending the primitive. something like `subscribe(or(as1, as2))` and `subscribe(and(as1, as2))` depending on if we want to abort if one of them is aborted or all of them.

Hmm, or maybe we need a "logic-gate like" abort source and subscribe to it normally.

I submitted https://github.com/scylladb/seastar/pull/1434 before seeing your reply, so let's move the discussion there for considering different extensions at the seastar layer.

avikivity commented:
I think we have a composability problem. Every feature needs support for every other feature. Complexity will explode and performance will tank.

bhalevy commented:
@avikivity can you please re-review?

kbr-scylla commented:
Is this new `send_message_cancellable` necessary? It's very similar to the existing `send_message_cancellable` that takes `abort_source&`. Which can also be used to implement timeouts (but only on the sending side) e.g. by aborting after some time passes (I used that in direct_failure_detector). Is there a difference, other than using `seastar::rpc::cancellable` instead of `abort_source`? Is the `cancellable` interface more flexible in some way? If yes, maybe we should remove the `abort_source` overload and replace existing usages with `cancellable`.

bhalevy commented:
> Is this new `send_message_cancellable` necessary? It's very similar to the existing `send_message_cancellable` that takes `abort_source&`. Which can also be used to implement timeouts (but only on the sending side) e.g. by aborting after some time passes (I used that in direct_failure_detector).

That was pretty much the approach in previous versions of this series, but @avikivity didn't like creating an ad-hoc abort_source and requesting its abort either via the long-living abort_source or in response to timeout, hence the use of `cancellable` which is a means for cancelling the message for various reasons.

>  Is there a difference, other than using `seastar::rpc::cancellable` instead of `abort_source`? Is the `cancellable` interface more flexible in some way? If yes, maybe we should remove the `abort_source` overload and replace existing usages with `cancellable`.

`cancellable` is indeed more flexible as the caller can decide to cancel for any reason, namely due to external abort (by tunneling an abort request using a a subscription to the external about_source) or a timeout (by setting up the timer callback).  The could be other reasons to cancel the outstanding message, like change of local state for example (e.g. loss of group0 leadership).

kbr-scylla commented:
> or a timeout (by setting up the timer callback). The could be other reasons to cancel the outstanding message, like change of local state for example (e.g. loss of group0 leadership).

We can also do that using `abort_source`, calling `request_abort()` on timer firing or whatever, so as I see it the flexibility is exactly the same. Yes you need an "ad-hoc" `abort_source` for that. But that isn't so different from creating an "ad_hoc" `rpc::cancellable` IMO. Is the `cancellable` interface easier to use?

In https://github.com/scylladb/scylladb/blob/a58beb8ce43b76d750fbfb6a9f6e154fbc62d03f/direct_failure_detector/failure_detector.cc#L470 I used the `abort_source` version of `send_message...` to implement timeouts (I didn't use the timeout version because the code is operating with abstract clocks).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17014/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17014/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8140/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16208/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4967/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17014/)
#### Build Details:
- Duration: 7 hr 9 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
@gleb-cloudius please re-review

gleb-cloudius commented on idl-compiler.py:
```diff
@@ -480,6 +480,9 @@ class RpcVerb(ASTBase):
       Incompatible with [[abortable]].
+    - [[cancellable]] - an additional rpc::cancellable& parameter is supplied
+      to the handler function and send* method uses send_message_*_cancellable
+      variant of internal function to actually send the message.
```
I do not understand why do you need cancellable and abortable. You need to make existing  cancellable and with_timeout attributes composable, not introduce this new version.

bhalevy replied:
today's `[[cancellable]]` expects reference to an abort_source which is really confusing, while the new `[[cancellable]]` expects reference to a `cancellable`, and so it is composable with `[[with_timeout]]`.
Making the existing `[[cancellable]]` composable with `[[with_timeout]]` will change the derived signature in a counterintuitive way.

bhalevy replied:
I think that to keep `[[cancellable]]` we can convert the existing code to use `cancellable` object rather than an abort_source and implement a shim to subscribe to the abort_source and signal the cancellable object.
I'll see if it's feasible.

bhalevy replied:
> > I think that to keep `[[cancellable]]` we can convert the existing code to use `cancellable` object rather than an abort_source and implement a shim to subscribe to the abort_source and signal the cancellable object. I'll see if it's feasible.
> 
> Why? Forget about `cancellable` . It is RPC implementation detail that messaging service is here to hide.

For performance sensitive RPCs, if the verb is called in a coroutine, the cancellable object can be put in the coroutine frame, saving an allocation that is done today to hook the `abort_source` to the `cancellable` object.

gleb-cloudius replied:
We can rewrite helpers with co-routines as well.

bhalevy commented:
In v6 ([4bd0845](https://github.com/scylladb/scylladb/commit/4bd0845fce71fe6b6a4a014f76917eb2590c59a8)):
- rebased
- simplify changes to idl-compiler, just make with_timeout and cancellable composable
  - and define `send_message_timeout_cancellable` accepting both a timeout and abort_source&

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17065/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17065/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5011/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8182/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16251/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17065/)
#### Failed Tests (1/43998):
- [test_data_resurrection_after_cleanup.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17065/testReport/junit/cluster/test_data_resurrection_after_cleanup/Tests___Unit_Tests___test_data_resurrection_after_cleanup_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_resurrection_after_cleanup.debug.2)
#### Build Details:
- Duration: 10 hr
- Builder: spider3.cloudius-systems.com

bhalevy commented:
> #### Failed Tests (1/43998):
> * [test_data_resurrection_after_cleanup.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17065/testReport/junit/cluster/test_data_resurrection_after_cleanup/Tests___Unit_Tests___test_data_resurrection_after_cleanup_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_data_resurrection_after_cleanup.debug.2)

@gleb-cloudius can you please investigate the failure above?
I wonder if the test assumptions are correct to decommission one out of two nodes.
Doesn't that cause loss of raft majority?
Should the test use a 3-node cluster instead?

gleb-cloudius commented:
> > #### Failed Tests (1/43998):
> > 
> > * [test_data_resurrection_after_cleanup.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17065/testReport/junit/cluster/test_data_resurrection_after_cleanup/Tests___Unit_Tests___test_data_resurrection_after_cleanup_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_data_resurrection_after_cleanup.debug.2)
> 
> @gleb-cloudius can you please investigate the failure above? I wonder if the test assumptions are correct to decommission one out of two nodes. Doesn't that cause loss of raft majority? Should the test use a 3-node cluster instead?

Decommissioning node removes it from the raft config, so only one suppose to be left in the cluster and operation should have a quorum. But we have a race where we report decommission completion before we remove the decommissioned node from the config, so if the decommissioned noticed that the operation is completed and exit before config change takes place we end up in a situation without a quorum.


```
case topology::transition_state::left_token_ring: {
...
co_await update_topology_state(take_guard(std::move(node)), {rtbuilder.build()}, "report request completion in left_token_ring state");
...
node = retake_node(co_await remove_from_group0(std::move(node.guard), node.id), node.id);
...
}
```

gleb-cloudius commented:
@bhalevy is there an issue about the failure?

bhalevy commented:
@scylladb/scylla-maint please consider merging
